source/common/redis/codec_impl.cc

```diff
@@ -0,0 +1,382 @@
+#include "codec_impl.h"
+
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+
+namespace Redis {
+
+std::vector<RespValue>& RespValue::asArray() {
+  ASSERT(type_ == RespType::Array);
+  return array_;
+}
+
+const std::vector<RespValue>& RespValue::asArray() const {
+  ASSERT(type_ == RespType::Array);
+  return array_;
+}
+
+std::string& RespValue::asString() {
+  ASSERT(type_ == RespType::BulkString || type_ == RespType::Error ||
+         type_ == RespType::SimpleString);
+  return string_;
+}
+
+const std::string& RespValue::asString() const {
+  ASSERT(type_ == RespType::BulkString || type_ == RespType::Error ||
+         type_ == RespType::SimpleString);
+  return string_;
+}
+
+int64_t& RespValue::asInteger() {
+  ASSERT(type_ == RespType::Integer);
+  return integer_;
+}
+
+int64_t RespValue::asInteger() const {
+  ASSERT(type_ == RespType::Integer);
+  return integer_;
+}
+
+void RespValue::cleanup() {
+  // Need to manually delete because of the union.
+  switch (type_) {
+  case RespType::Array: {
+    array_.~vector<RespValue>();
+    break;
+  }
+  case RespType::SimpleString:
+  case RespType::BulkString:
+  case RespType::Error: {
+    string_.~basic_string<char>();
+    break;
+  }
+  case RespType::Null:
+  case RespType::Integer: {
+    break;
+  }
+  }
+}
+
+void RespValue::type(RespType type) {
+  cleanup();
+
+  // Need to use placement new because of the union.
+  type_ = type;
+  switch (type) {
+  case RespType::Array: {
+    new (&array_) std::vector<RespValue>();
+    break;
+  }
+  case RespType::SimpleString:
+  case RespType::BulkString:
+  case RespType::Error: {
+    new (&string_) std::string();
+    break;
+  }
+  case RespType::Null:
+  case RespType::Integer: {
+    break;
+  }
+  }
+}
+
+void DecoderImpl::decode(Buffer::Instance& data) {
+  uint64_t num_slices = data.getRawSlices(nullptr, 0);
+  Buffer::RawSlice slices[num_slices];
+  data.getRawSlices(slices, num_slices);
+  for (Buffer::RawSlice& slice : slices) {
+    parseSlice(slice);
+  }
+
+  data.drain(data.length());
+}
+
+void DecoderImpl::parseSlice(const Buffer::RawSlice& slice) {
+  char* buffer = reinterpret_cast<char*>(slice.mem_);
+  uint64_t remaining = slice.len_;
+
+  while (remaining || state_ == State::ValueComplete) {
```

Comment:
Might refactor later. The code is easy to read w/ switch statement so will keep for now.